proxy.go

```diff
@@ -162,12 +162,87 @@ func (rp *reverseProxy) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
 	requestDuration.With(s.labels).Observe(since)
 }
 
+func executeWithRetry(
+	ctx context.Context,
+	s *scope,
+	startTime time.Time,
+	retryNum int,
+	rp func(http.ResponseWriter, *http.Request),
+	rw ResponseWriterWithCode,
+	srw *statResponseWriter,
+	req *http.Request,
+	proxiedResponseDuration *prometheus.SummaryVec,
+) error {
+	numReplicas := len(s.host.replica.cluster.replicas)
+	if numReplicas > 1 && retryNum <= numReplicas {
+		for i := 0; i <= retryNum; i++ {
+			rp(rw, req)
+
+			err := ctx.Err()
+			if err == nil { //nolint: gocritic
+				// The request has been successfully proxied.
+
+				since := time.Since(startTime).Seconds()
+				proxiedResponseDuration.With(s.labels).Observe(since)
+
+				srw.statusCode = rw.StatusCode()
+
+				// StatusBadGateway response is returned by http.ReverseProxy when
+				// it cannot establish connection to remote host.
+				if rw.StatusCode() == http.StatusBadGateway {
+					if i == retryNum {
+						s.host.penalize()
+						q := getQuerySnippet(req)
+						err1 := fmt.Errorf("%s: cannot reach %s; query: %q", s, s.host.addr.Host, q)
+						respondWith(srw, err1, srw.statusCode)
+					} else {
+						h := s.host
+						h.dec()
+						h.active = 0
+						s.host = h.replica.getHost()
+						h.active = 1
+						req.URL.Host = s.host.addr.Host
+						req.URL.Scheme = s.host.addr.Scheme
+					}
+				} else {
+					return nil
+				}
+			} else {
+				return err
+			}
```

Comment:
```suggestion
			if err != nil {
			     return err
			}

				since := time.Since(startTime).Seconds()
				proxiedResponseDuration.With(s.labels).Observe(since)

				srw.statusCode = rw.StatusCode()

				// StatusBadGateway response is returned by http.ReverseProxy when
				// it cannot establish connection to remote host.
				if rw.StatusCode() == http.StatusBadGateway {
					if i == retryNum {
						s.host.penalize()
						q := getQuerySnippet(req)
						err1 := fmt.Errorf("%s: cannot reach %s; query: %q", s, s.host.addr.Host, q)
						respondWith(srw, err1, srw.statusCode)
					} else {
						h := s.host
						h.dec()
						h.active = 0
						s.host = h.replica.getHost()
						h.active = 1
						req.URL.Host = s.host.addr.Host
						req.URL.Scheme = s.host.addr.Scheme
					}
				} 
					return nil
```

Reply:
I made a change here, the return nil  should under else condition.  In order to continue the for loop. 
```suggestion
		for i := 0; i <= retryNum; i++ {
			rp(rw, req)

			err := ctx.Err()
			if err != nil {
				return err
			}
			since := time.Since(startTime).Seconds()
			proxiedResponseDuration.With(s.labels).Observe(since)

			srw.statusCode = rw.StatusCode()
			if rw.StatusCode() == http.StatusBadGateway {
				if i == retryNum {
					s.host.penalize()
					q := getQuerySnippet(req)
					err1 := fmt.Errorf("%s: cannot reach %s; query: %q", s, s.host.addr.Host, q)
					respondWith(srw, err1, srw.statusCode)
				} else {
					h := s.host
					h.dec()
					h.active = 0
					s.host = h.replica.getHost()
					h.active = 1
					req.URL.Host = s.host.addr.Host
					req.URL.Scheme = s.host.addr.Scheme
				}
			} else {
				return nil
			}
		}
```